From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Tue, 8 Feb 2022 15:22:39 +0100
Subject: openaptx: Support libfreeaptx

[libfreeaptx][1] is a fork of libopenapt 0.2.0, used by pipewire.

[1]: https://github.com/iamthehorker/libfreeaptx

Fixes: https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/issues/1642
Closes: https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/issues/1589
Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/1667>
---
 ext/openaptx/gstopenaptxdec.h |  4 ++++
 ext/openaptx/gstopenaptxenc.h |  4 ++++
 ext/openaptx/meson.build      | 15 +++++++++++++--
 3 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/ext/openaptx/gstopenaptxdec.h b/ext/openaptx/gstopenaptxdec.h
index 95b126c..081b194 100644
--- a/ext/openaptx/gstopenaptxdec.h
+++ b/ext/openaptx/gstopenaptxdec.h
@@ -24,7 +24,11 @@
 #include <gst/gst.h>
 #include <gst/audio/audio.h>
 
+#ifdef USE_FREEAPTX
+#include <freeaptx.h>
+#else
 #include <openaptx.h>
+#endif
 
 G_BEGIN_DECLS
 
diff --git a/ext/openaptx/gstopenaptxenc.h b/ext/openaptx/gstopenaptxenc.h
index ab80b61..4cf7e17 100644
--- a/ext/openaptx/gstopenaptxenc.h
+++ b/ext/openaptx/gstopenaptxenc.h
@@ -24,7 +24,11 @@
 #include <gst/gst.h>
 #include <gst/audio/audio.h>
 
+#ifdef USE_FREEAPTX
+#include <freeaptx.h>
+#else
 #include <openaptx.h>
+#endif
 
 G_BEGIN_DECLS
 
diff --git a/ext/openaptx/meson.build b/ext/openaptx/meson.build
index e361ef3..4db8fdb 100644
--- a/ext/openaptx/meson.build
+++ b/ext/openaptx/meson.build
@@ -4,12 +4,23 @@ openaptx_sources = [
   'gstopenaptxenc.c',
 ]
 
-openaptx_dep = dependency('libopenaptx', version : '== 0.2.0', required : get_option('openaptx'))
+if not get_option('openaptx').allowed()
+  subdir_done()
+endif
+
+openaptx_defines = []
+
+openaptx_dep = dependency('libfreeaptx', version : '>= 0.1.1', required : false)
+if openaptx_dep.found()
+  openaptx_defines += ['-DUSE_FREEAPTX']
+else
+  openaptx_dep = dependency('libopenaptx', version : '== 0.2.0', required : get_option('openaptx'))
+endif
 
 if openaptx_dep.found()
   gstopenaptx = library('gstopenaptx',
     openaptx_sources,
-    c_args : gst_plugins_bad_args,
+    c_args : gst_plugins_bad_args + openaptx_defines,
     include_directories : [configinc],
     dependencies : [gstaudio_dep, openaptx_dep],
     install : true,
